Cleanup flash enablement for NXP Platforms

danieldegrasse opened the pull request:
This PR is an effort to cleanup flash enablement for all NXP platforms. As part of the work in #48154, a bug was exposed with NXP's platforms. We typically do not explicitly set the flash controller to disabled in the SOC devicetree, and due to this flash drivers were being enabled on SOCs they have not been tested with. This PR attempts to cleanup and document support for flash on NXP platforms.

### Methodology
Given that prior to #48154, specific KConfigs were used to indicate support for a flash driver, devices to target for testing were based on what was supported with the 3.1 release, with the addition of platforms I know were enabled since then.

### Changes
- For all SOCs verified to work:
   - the flash controller is now explicitly marked as enabled at the SOC level (for on chip flash) or board level (for external flash)
   - flash partition definitions are present at the board level.
   - a new `flash-ctlr` definition was added to the board's YAML, and a flash controller entry was added to the board's supported peripheral list
- For all SOCs not currently supported, the flash controller is explicitly marked as "disabled" in the SOC devicetree
- The flash driver test write size was increased to accommodate the LPC flash controller's requirement of minimum write lengths of 512 bytes.
- The flash driver test (`tests/drivers/flash`) has been updated to use the `flash-ctlr` YAML definition, when present, to enable the flash driver test.

### Testing
Platforms were verified to pass `tests/drivers/flash`. Other samples were not tested, primarily due to the 512 byte alignment and write length restrictions of LPC flash controllers.

### Results
The following boards were verified to work with this PR using the testing methodology described above: 

- mimxrt1050_evk
- lpcxpresso54114_m4
- lpcxpresso55s06
- lpcxpresso55s16
- lpcxpresso55s69_cpu0
- lpcxpresso55s28
- frdm_k82f
- frdm_kw41z
- frdm_k64f
- frdm_k22f
- mimxrt595_evk_cm33
- mimxrt685_evk_cm33
- mimxt1060_evk
- mimxrt1020_evk
- mimxrt1170_evk_cm4
- mimxrt1170_evk_cm7
- mimxrt1160_evk_cm7
- mimxrt1160_evk_cm4
- mimxrt1024_evk
- mimxrt1064_evk
- mimxrt1050_evk_qspi
- usb_kw24d512

The following boards are enabled, but could not be tested (as I don't have the hardware)
- twr_kv58f220m
- twr_ke18f
- mimxrt1060_evk_hyperflash

The following boards are not supported with flash samples:
- lpcxpresso11u68
- lpcxpresso55s69_cpu1
- lpcxpresso55s69_ns
- lpcxpresso54114_m0
- frdm_kl25z
- mimx8mm_evk
- mimxrt1015_evk
- mimxrt1010_evk
- mimx8mq_evk_cm4
- mimx8mp_evk_ddr
- mimx8mp_evk_itcm
- lpcxpresso51u68


### Linked Issues
Fixes #49073 
Fixes #49166

danieldegrasse committed:
soc: nxp_rt6xx: disable flash driver by default

set nxp,imx-flexspi to disabled by default to disable flash driver for
this SOC unless enabled by the target board. This SOC uses external flash,
so boards should enable the flash controller only when flash support
has been verified with their specific external flash module

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt685_evk_cm33: enable flash

enable flash node for mimxrt685_evk_cm33

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1050_evk: document flexspi support

Document support for flexspi flash controller on mimxrt1050_evk and
mimxrt1050_evk_qspi targets

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1060_evk: document flexspi support

Document support for using flexspi as a flash controller on
mixmrt1060_evk and mimxrt1060_evkb

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1170_evk: document flexspi support

Document support for using flexspi as a flash controller on
mimxrt1170_evk_cm7 and mimxrt1170_evk_cm4

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1020_evk: document flexspi support

Document support for flexspi as a flash controller on mimxrt1020_evk

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
tests: drivers: flash: increase write size of test

Some SOCs, such as NXP's LPC family, expect a minimum write size of 512
bytes. Increase the EXPECTED_SIZE definition for flash driver test to
512 so that test can pass on these platforms.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_lpc55s6x: disable flash controller for nonsecure targets

disable flash controller for nonsecure core, as only the secure core is
tested with flash support.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s69: update documentation to indicate IAP support

Update documentation to indicate that lpcxpresso55s69_cpu0 supports IAP
flash controller.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_k6x: enable flash controller ftfe

explicitly enable ftfe flash controller for k6x SOC, as it is tested and
supported.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_k64f: document flash controller support

document support for FTFE flash controller

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_k2x: enable ftfe flash controller

Mark ftfe flash controller as enabled, as it is tested on the k2x SOC line

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_k22f: document support for flash controller

Document support for FTFE flash controller

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_k8x.dtsi: enable ftfa flash controller

Enable flash controller on k8x, as it is supported on this SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_k82f: document support for ftfa flash controller

Document support for ftfa flash controller on frdm_k82f

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1064_evk: document flexspi support

Document support for flexspi as flash controller on mimxrt1064_evk, and
move status=okay for internal flash to soc level dtsi

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1024_evk: Update flexspi partition layout and device name

Update flexspi partition layout to add a 128KB storage partition. Also,
fix flash device name and size to match the SIP flash present on
the RT1024 SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_rt: Remove flexspi2 on RT SOCs that don't have it

Remove flexspi2 node for NXP's iMX.RT SOCs that only have one flexspi
controller

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1160_evk: update documentation to indicate flexspi support

Update board documentation to indicate support for using flexspi as
a flash controller device

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: flash: soc_flash_lpc: enable support for lpc54xxx IAP

Add support for LPC54xxx IAP flash driver to soc_flash_lpc.c
Driver is tested on M4 core only, and is therefore disabled on the M0 core.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso54114: enable iap flash controller

Enable flash driver with IAP flash controller

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_lpc55S0x: disable IAP by default

Disable IAP by default for lpc55s06, enable it for the non secure core
as this is the mode that the flash controller has been tested in.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s06: enable IAP flash controller

enable IAP flash controller for lpcxpresso55s06 evk

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_kw41z: enable flash controller

explicitly enable flash controller, as it is supported on this SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_kw41z: enable flash controller

Enable flash controller for frdm_kw41z evk

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_lpc55S2x: enable IAP for secure core

enable IAP flash controller for secure core, as this is the core where
flash support has been verified.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s28: enable IAP flash controller

enable IAP flash controller for lpcxpresso55s28 evk

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_lpc55S1x: enable IAP for non secure core

enable IAP for non secure core, as it is tested on this core only

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s16: enable IAP flash controller

enable IAP flash controller for lpcxpresso55s16 evk

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_rt5xx_common: disable flexspi by default

Disable flexspi flash controller by default, allow boards to enable it.
This SOC uses external flash, so boards should only enable the flash
controller when their specific flash module has been verified to work
with the flash driver API in Zephyr.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt595_evk_cm33: enable flexspi as flash controller

enable using flexspi as flash controller for RT595 EVK

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_kl25z: disable flash controller

Disable flash controller for kl25z SOC, as flash driver support is
not tested on this SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_kw40z: disable ftfa flash controller

Disable ftfa flash controller, as flash driver support is not tested
for this SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: usb_kw24d512: enable flash controller

enable flash controller, as it is tested on this SOC

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: usb_kw24d512: enable flash controller

enable flash controller on usb_kw24d512d

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

galak reviewed (commented):
What add `flash_ctlr` to board.yaml's?  We don't use that anywhere

galak commented on dts/arm/nxp/nxp_kl25z.dtsi:
```diff
@@ -39,6 +39,7 @@
 			compatible = "nxp,kinetis-ftfa";
 			reg = <0x40020000 0x14>;
 			interrupts = <5 0>;
+			status = "disabled";
```
Why is this disabled?  (at least add comment in commit message)

danieldegrasse replied:
Will update commit message, but the flash controller is disabled on this platform because it's not currently tested to work on the kl25z soc.

galak commented on dts/arm/nxp/nxp_kw40z.dtsi:
```diff
@@ -78,6 +78,7 @@
 			compatible = "nxp,kinetis-ftfa";
 			reg = <0x40020000 0x2c>;
 			interrupts = <5 0>;
+			status = "disabled";
```
why is this disabled, at least add comment in commit message.

danieldegrasse replied:
Same reason as above, will update commit message to reflect this

galak commented on dts/arm/nxp/nxp_rt5xx_common.dtsi:
```diff
@@ -455,6 +455,7 @@
 
 &flexspi {
 	compatible = "nxp,imx-flexspi";
+	status = "disabled";
```
why is this disabled, at least add comment in commit message.  Something like 'flexspi will be board specific modules of FLASH so leave it to the board to enable...'

danieldegrasse replied:
Updated commit message, as you pointed out this is disabled at the SOC level because the external flash module is board specific and needs to be tested on a per board basis.

galak commented on dts/arm/nxp/nxp_rt6xx_common.dtsi:
```diff
@@ -406,6 +406,7 @@
 	interrupts = <42 0>;
 	#address-cells = <1>;
 	#size-cells = <0>;
+	status = "disabled";
```
why is this disabled, at least add comment in commit message.  Something like 'flexspi will be board specific modules of FLASH so leave it to the board to enable...'

galak reviewed (changes requested):
Some commit comments need updating to explain why in a few cases.

danieldegrasse commented:
> What add flash_ctlr to board.yaml's? We don't use that anywhere

Reasoning here is to start providing a yaml binding that will mark flash controllers as supported on specific boards, same way we indicate support for SPI or I2C. I'll update the flash driver testcase to use this yaml.

danieldegrasse committed:
tests: drivers: flash: Make testcase depend on FLASH_HAS_DRIVER_ENABLED

Change default flash driver testcase to depend on FLASH_HAS_DRIVER_ENABLED,
and the presence of a storage partition in the output DTS, so that the test
will run on all platforms that report flash support.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

galak commented:
> > 
> 
> Reasoning here is to start providing a yaml binding that will mark flash controllers as supported on specific boards, same way we indicate support for SPI or I2C. I'll update the flash driver testcase to use this yaml.

So no other platforms do this, I don't think we should start adding a new `flash_ctrl` tag now.

danieldegrasse committed:
drivers: flash: flash_mcux_flexspi_hyperflash: Fix incorrect printf format

Use correct printf format specifier for LOG_DBG calls using offsets, as
these offsets are long int and thus require the %lx format specifier rather
than %x.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

galak commented:
> I chose the YAML tag because the flash driver testcase requires a storage partition be defined, which is likely not present on every board in Zephyr that supports flash controllers. The `FLASH_HAS_DRIVER_ENABLED` filtering would enable this test for boards that likely won't support it. As far as filtering using the dt_compatible, as I'm sure you are aware NXP has quite a few devicetree bindings for flash, and I'd prefer to avoid filling this testcase with them and instead use a YAML tag at the board level

The YAML tag does not imply partitions being defined at all.  Why can't you se a combo of `FLASH_HAS_DRIVER_ENABLED` and dt filter for partition being defined.

danieldegrasse commented:
> The YAML tag does not imply partitions being defined at all. Why can't you use a combo of FLASH_HAS_DRIVER_ENABLED and dt filter for partition being defined.

Good point about the implications of the YAML tag. I've updated this PR to use `FLASH_HAS_DRIVER_ENABLED` with the dt filter.

galak commented on tests/drivers/flash/testcase.yaml:
```diff
@@ -27,6 +27,11 @@ tests:
+    tags: drivers flash
+    filter: CONFIG_FLASH_HAS_DRIVER_ENABLED and
+        dt_label_with_parent_compat_enabled("storage_partition", "fixed-partitions")
+    depends_on: flash_ctlr
```
Need to remove the `depends_on`

danieldegrasse replied:
Fixed, sorry I missed that

danieldegrasse committed:
tests: drivers: flash: enable heap for ESP32 boards

Enable heap for ESP32 boards that require it for flash driver support.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
